models/git/protected_branch_list.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package git
+
+import (
+	"context"
+	"sort"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/git"
+
+	"github.com/gobwas/glob"
+)
+
+type ProtectedBranchRules []*ProtectedBranch
+
+func (rules ProtectedBranchRules) GetFirstMatched(branchName string) *ProtectedBranch {
```

Comment:
maybe store the `rule-branch` relate result in database by a new table?

Reply:
Why?

Reply:
Hmm, maybe can simplify check logic. will only check it when add new rule or new branch. and get the result for a branch from database when need use it.

Reply:
But we need to store all branches and sync them with git at first, os leave it for other PRs.

Reply:
> But we need to store all branches and sync them with git at first, os leave it for other PRs.

Hmm, maybe not needed extern rule-branch record for the PlainRule which only contain one branch. can get it directly like previous logic.